sorts/heapsort.ts

```diff
@@ -0,0 +1,46 @@
+/* 
+Implementation of in-place heap sort.
+
+  complexity:
+    best time - O(N log N)
+    avg time - O(N log N)
+    worst time - O(N log N)
+
+    memory - O(1)
+    in-place - Yes
+    stable - No
+*/
+
+module.exports = function HeapSort(arr: number[]): number[] {
+  const left = (i: number): number => i * 2 + 1,
+    right = (i: number): number => i * 2 + 2;
+
+  function heapifyBottom(i: number, limit = arr.length): void {
+    while (1) {
+      // pretend if we try to acces to element below limit it's undefined
+      const l = left(i) < limit ? left(i) : arr.length,
+        r = right(i) < limit ? right(i) : arr.length;
+      if (arr[r] > arr[i] && !(arr[r] < arr[l])) {
```

Comment:
I'd flip the condition: `arr[r] >= arr[l]`

Reply:
Ye, first one can be obviously flipped, but second one is have to be extended, because it uses `undefined` propertied in comparisons

Reply:
So, we can change 
`} else if (arr[l] > arr[i] && !(arr[l] < arr[r])) {`
to
` } else if (arr[l] > arr[i] && (arr[l] >= arr[r] || arr[r] === undefined)) {`
but my text formatter think it's too long and make it into 4 lines, so i will write this way:
` } else if (arr[l] > arr[i] && (arr[l] >= arr[r] || !(r in arr))) {`